p4-16/spec/P4-16-spec.mdk

```diff
@@ -6769,55 +6782,81 @@ and illegal specifications of default actions for the `table t`:
 
 While table entries are typically installed by the control plane,
 tables may also be initialized at compile time with a set of entries.
-This is useful in situations where tables are
+
+Declaring these entries with `const entries`
+is useful in situations where tables are
 used to implement fixed algorithms---defining table entries statically
 enables expressing these algorithms directly in P4,
 which allows the compiler to infer how the table is actually used and
 potentially make better allocation decisions for targets with limited
-resources. Entries declared in the P4 source are installed in the
+resources.
+
+Declaring entries with `entries` (without the `const` qualifier)
+enables one to specify a mix of some immutable entries that are always
+in the table, and some mutable entries that the control plane is
+allowed to later change or remove.
+
+Entries declared in the P4 source are installed in the
 table when the program is loaded onto the target. Entries cannot be
 specified for a table with no key (see Sec. [#sec-table-keys]).
 
 Table entries are defined using the following syntax:
 
 ~ Begin P4Grammar
 tableProperty
- : const ENTRIES '=' '{' entriesLlist '}' /* immutable entries */
+ : optAnnotations const ENTRIES '=' '{' entriesLlist '}' /* immutable entries */
+ | optAnnotations ENTRIES '=' '{' entriesList '}' /* mix of mutable and/or immutable entries */
 
 entriesList
  : /* empty */
  | entriesList entry
  ;
 
+optCONST
+ : /* empty */
+ | const
+ ;
+
+optEntryPriority
+ : /* empty */
+ | priority '=' INTEGER
+ ;
+
 entry
- : keysetExpression ':' actionRef optAnnotations ';'
+ : optCONST optEntryPriority keysetExpression ':' actionRef optAnnotations ';'
  ;
 ~ End P4Grammar
 
 Table entries defined using `const entries` are immutable---i.e., they
 can only be read by the control plane.  The control plane is not
 allowed to remove or modify any entries defined within `const
 entries`, nor is it allowed to add entries to such a table.
+It is allowed for individual entries to have the `const` keyword
+before them, but this is redundant when the entries are declared using
+`const entries`.
+
+Table entries defined using `entries` (without a `const` qualifier
+before it) may have `const` before them, or not, independently for
+each entry.  Entries with `const` before them may not be modified or
+removed by the control plane.  Entries without `const` may be modified
+or removed by the control plane.  It is permitted for the control
+plane to add entries to such a table (subject to table capacity
+limitations), unlike tables declared with `const entries`.
```

Comment:
I don't know if it's worth specifying that having an entry in a table that is unreachable is possible, so the fact something can't be removed doesn't mean it will ever apply.

Reply:
By "unreachable" do you mean that it is impossible that a table.apply() call could ever find it as the "winning entry", because other higher priority entries mask it?

If so, this concept is never mentioned in the spec, and as far as I can see, need not be mentioned in the spec.

If an implementation wants to optimize away such entries, that seems to me a fine optimization that an implementation could make, but the spec doesn't need to mandate or forbid it.  The spec defines the observable behavior.